OpenRA.Game/Graphics/SpriteLoader.cs

```diff
@@ -56,21 +58,57 @@ public SpriteCache(IReadOnlyFileSystem fileSystem, ISpriteLoader[] loaders, Shee
 			this.loaders = loaders;
 		}
 
-		Sprite[] LoadSprite(string filename, List<Sprite[]> cache)
-		{
-			var sprite = SpriteLoader.GetSprites(fileSystem, filename, loaders, SheetBuilder);
-			cache.Add(sprite);
-			return sprite;
-		}
-
-		/// <summary>Returns the first set of sprites with the given filename.</summary>
-		public Sprite[] this[string filename]
+		/// <summary>
+		/// Returns the first set of sprites with the given filename.
+		/// If getUsedFrames is defined then the indices returned by the function call
+		/// are guaranteed to be loaded.  The value of other indices in the returned
+		/// array are undefined and should never be accessed.
+		/// </summary>
+		public Sprite[] this[string filename, Func<int, IEnumerable<int>> getUsedFrames = null]
 		{
 			get
 			{
 				var allSprites = sprites.GetOrAdd(filename);
 				var sprite = allSprites.FirstOrDefault();
-				return sprite ?? LoadSprite(filename, allSprites);
+
+				ISpriteFrame[] unloaded;
+				if (!unloadedFrames.TryGetValue(filename, out unloaded))
+					unloaded = null;
+
+				// This is the first time that the file has been requested
+				// Load all of the frames into the unused buffer and initialize
+				// the loaded cache (initially empty)
+				if (sprite == null)
+				{
+					unloaded = FrameLoader.GetFrames(fileSystem, filename, loaders);
+					unloadedFrames[filename] = unloaded;
+
+					sprite = new Sprite[unloaded.Length];
+					allSprites.Add(sprite);
+				}
+
+				// HACK: The sequency code relies on side-effects from getUsedFrames
```

Comment:
Supporting `Length: *` in the sequence yaml puts us in the ugly recursive situation where the yaml metadata depends on the sprite data which depends on the yaml data.  This is not the first time that this has bit us.  Our sequence parsing code would simplify a lot if we could drop that and require all lengths to be explicitly defined.